File: back/src/main/java/com/triptrace/domain/trip/trip/service/TripService.java
package com.triptrace.domain.trip.trip.service;

import com.triptrace.domain.image.image.entity.Image;
import com.triptrace.domain.image.image.repository.ImageRepository;
import com.triptrace.domain.marker.marker.repository.MarkerRepository;
import com.triptrace.domain.member.member.entity.Member;
import com.triptrace.domain.member.member.repository.MemberRepository;
import com.triptrace.domain.post.post.entity.Post;
import com.triptrace.domain.post.post.repository.PostRepository;
import com.triptrace.domain.trip.trip.dto.TripCreateRequest;
import com.triptrace.domain.trip.trip.dto.TripModifyRequest;
import com.triptrace.domain.trip.trip.dto.TripResponse;
import com.triptrace.domain.trip.trip.entity.Trip;
import com.triptrace.domain.trip.trip.error.TripErrorCode;
import com.triptrace.domain.trip.trip.repository.TripRepository;
import com.triptrace.domain.trip.tripLike.repository.TripLikeRepository;
import com.triptrace.global.exception.ServiceException;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.time.LocalDateTime;
import java.util.List;

@Service
public class TripService {
    private final TripRepository tripRepository;
    private final MemberRepository memberRepository;
    private final ImageRepository imageRepository;
    private final PostRepository postRepository;
    private final MarkerRepository markerRepository;
    private final TripLikeRepository tripLikeRepository;

    @Transactional
    public TripResponse create(Long ownerId, TripCreateRequest request) {
        Member owner = memberRepository.findById(ownerId)
            .orElseThrow(() -> new ServiceException(TripErrorCode.MEMBER_NOT_FOUND));
        Trip trip = tripRepository.save(new Trip(
            owner,
            request.title(),
            request.country(),
            request.city(),
            request.startDate(),
            request.endDate(),
            request.visibility()
        ));

        return toResponse(trip);
    }

    @Transactional(readOnly = true)
    public List<TripResponse> findTripsByOwnerId(Long ownerId) {
        return tripRepository.findByOwnerId(ownerId)
            .stream()
            .map(this::toResponse)
            .toList();
    }

    @Transactional(readOnly = true)
    public Page<TripResponse> findTripsByOwnerId(Long ownerId, Pageable pageable) {
        return tripRepository.findByOwnerIdOrderByCreatedAtDescIdDesc(ownerId, pageable)
            .map(this::toResponse);
    }

    @Transactional(readOnly = true)
    public List<TripResponse> findPublicTrips() {
        return tripRepository.findByVisibilityTrue()
            .stream()
            .map(this::toResponse)
            .toList();
    }

    @Transactional(readOnly = true)
    public TripResponse findAccessibleTrip(Long tripId, Long ownerId) {
        Trip trip = tripRepository.findById(tripId)
            .orElseThrow(() -> new ServiceException(TripErrorCode.NOT_FOUND));

        if (!trip.isVisibility()) {
            validateOwner(trip, ownerId);
        }

        return toResponse(trip);
    }

    @Transactional(readOnly = true)
    public Trip findOwnedTrip(Long tripId, Long ownerId) {
        Trip trip = tripRepository.findById(tripId)
            .orElseThrow(() -> new ServiceException(TripErrorCode.NOT_FOUND));
        // 이미지 업로드처럼 여행기 내부 데이터를 변경하는 로직은 공개 여부와 무관하게 소유자만 접근 가능
        validateOwner(trip, ownerId);

        return trip;
    }

    @Transactional
    public TripResponse modifyTrip(Long tripId, Long ownerId, TripModifyRequest request) {
        Trip trip = findOwnedTrip(tripId, ownerId);

        trip.modify(
            request.title(),
            request.country(),
            request.city(),
            request.startDate(),
            request.endDate(),
            request.visibility()
        );

        return toResponse(trip);
    }

    @Transactional
    public void deleteTrip(Long tripId, Long ownerId) {
        Trip trip = findOwnedTrip(tripId, ownerId);
        List<Post> posts = postRepository.findByTripId(tripId);
        List<Long> postIds = posts.stream()
            .map(Post::getId)
            .toList();

        trip.changeRepresentativeImage(null);

        if (!postIds.isEmpty()) {
            markerRepository.deleteAll(markerRepository.findByPostIdIn(postIds));
        }

        tripLikeRepository.deleteByTripId(tripId);
        imageRepository.deleteAll(imageRepository.findByTripId(tripId));
        postRepository.deleteAll(posts);
        tripRepository.delete(trip);
    }

    @Transactional
    public TripResponse changeRepresentativeImage(Long tripId, Long ownerId, Long imageId) {
        Trip trip = findOwnedTrip(tripId, ownerId);
        Image image = imageRepository.findById(imageId)
            .orElseThrow(() -> new ServiceException(TripErrorCode.IMAGE_NOT_FOUND));

        if (!image.getTrip().getId().equals(trip.getId()) || !image.getOwner().getId().equals(ownerId)) {
            throw new ServiceException(TripErrorCode.IMAGE_FORBIDDEN);
        }
        trip.changeRepresentativeImage(image);
        return toResponse(trip);
    }

    // 공개여행기 중 최근 한 달 좋아요 수 상위 10개 조회
    @Transactional(readOnly = true)
    public List<TripResponse> findTop10PublicTripsByLikeCount() {
        LocalDateTime likedSince = LocalDateTime.now().minusMonths(1);

        return tripRepository.findTop10PublicTripsByRecentLikeCount(likedSince)
            .stream()
            .map(this::toResponse)
            .toList();
    }

    // 공개여행기 중 createdAt 기준 내림차순 조회 메서드 추가
    @Transactional(readOnly = true)
    public List<TripResponse> findPublicTripsByCreatedAtDesc() {
        return tripRepository.findByVisibilityTrueOrderByCreatedAtDesc()
            .stream()
            .map(this::toResponse)
            .toList();
    }

    @Transactional(readOnly = true)
    public Page<TripResponse> findPublicTripsByCreatedAtDesc(Pageable pageable) {
        return tripRepository.findByVisibilityTrueOrderByCreatedAtDescIdDesc(pageable)
            .map(this::toResponse);
    }

    private TripResponse toResponse(Trip trip) {
        return new TripResponse(trip);
    }

    private void validateOwner(Trip trip, Long ownerId) {
        if (!trip.getOwner().getId().equals(ownerId)) {
            throw new ServiceException(TripErrorCode.FORBIDDEN);
        }
    }

    public TripService(final TripRepository tripRepository, final MemberRepository memberRepository, final ImageRepository imageRepository, final PostRepository postRepository, final MarkerRepository markerRepository, final TripLikeRepository tripLikeRepository) {
        this.tripRepository = tripRepository;
        this.memberRepository = memberRepository;
        this.imageRepository = imageRepository;
        this.postRepository = postRepository;
        this.markerRepository = markerRepository;
        this.tripLikeRepository = tripLikeRepository;
    }
}

File: back/src/main/kotlin/com/triptrace/domain/trip/trip/service/TripService.kt
package com.triptrace.domain.trip.trip.service

import com.triptrace.domain.image.image.repository.ImageRepository
import com.triptrace.domain.marker.marker.repository.MarkerRepository
import com.triptrace.domain.member.member.repository.MemberRepository
import com.triptrace.domain.post.post.repository.PostRepository
import com.triptrace.domain.trip.trip.dto.PopularDestinationResponse
import com.triptrace.domain.trip.trip.dto.TripCreateRequest
import com.triptrace.domain.trip.trip.dto.TripModifyRequest
import com.triptrace.domain.trip.trip.dto.TripResponse
import com.triptrace.domain.trip.trip.dto.WeeklyTrendingTripResponse
import com.triptrace.domain.trip.trip.entity.Trip
import com.triptrace.domain.trip.trip.error.TripErrorCode
import com.triptrace.domain.trip.trip.repository.TripRepository
import com.triptrace.domain.trip.tripLike.repository.TripLikeRepository
import com.triptrace.global.exception.ServiceException
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime

@Service
class TripService(
    private val tripRepository: TripRepository,
    private val memberRepository: MemberRepository,
    private val imageRepository: ImageRepository,
    private val postRepository: PostRepository,
    private val markerRepository: MarkerRepository,
    private val tripLikeRepository: TripLikeRepository,
) {
    @Transactional fun create(
        ownerId: Long,
        request: TripCreateRequest,
    ): TripResponse {
        val owner = memberRepository.findById(ownerId).orElseThrow { ServiceException(TripErrorCode.MEMBER_NOT_FOUND) }
        return TripResponse(
            tripRepository.save(
                Trip(owner, request.title, request.country, request.city, request.startDate, request.endDate, request.visibility),
            ),
        )
    }

    @Transactional(readOnly = true)
    fun findTripsByOwnerId(ownerId: Long): List<TripResponse> = tripRepository.findByOwnerId(ownerId).map(::TripResponse)

    @Transactional(readOnly = true)
    fun findTripsByOwnerId(
        ownerId: Long,
        pageable: Pageable,
    ): Page<TripResponse> = tripRepository.findByOwnerIdOrderByCreatedAtDescIdDesc(ownerId, pageable).map(::TripResponse)

    @Transactional(readOnly = true)
    fun findPublicTrips(): List<TripResponse> = tripRepository.findByVisibilityTrue().map(::TripResponse)

    @Transactional(readOnly = true)
    fun findAccessibleTrip(
        tripId: Long,
        ownerId: Long?,
    ): TripResponse {
        val trip = findTrip(tripId)
        if (!trip.visibility) validateOwner(trip, ownerId)
        return TripResponse(trip)
    }

    @Transactional(readOnly = true)
    fun findOwnedTrip(
        tripId: Long,
        ownerId: Long,
    ): Trip = findTrip(tripId).also { validateOwner(it, ownerId) }

    @Transactional fun modifyTrip(
        tripId: Long,
        ownerId: Long,
        request: TripModifyRequest,
    ): TripResponse {
        val trip = findOwnedTrip(tripId, ownerId)
        trip.modify(request.title, request.country, request.city, request.startDate, request.endDate, request.visibility)
        return TripResponse(trip)
    }

    @Transactional fun deleteTrip(
        tripId: Long,
        ownerId: Long,
    ) {
        val trip = findOwnedTrip(tripId, ownerId)
        val posts = postRepository.findByTripId(tripId)
        val postIds = posts.mapNotNull { it.id }
        trip.changeRepresentativeImage(null)
        if (postIds.isNotEmpty()) markerRepository.deleteAll(markerRepository.findByPostIdIn(postIds))
        tripLikeRepository.deleteByTripId(tripId)
        imageRepository.deleteAll(imageRepository.findByTripId(tripId))
        postRepository.deleteAll(posts)
        tripRepository.delete(trip)
    }

    @Transactional fun changeRepresentativeImage(
        tripId: Long,
        ownerId: Long,
        imageId: Long,
    ): TripResponse {
        val trip = findOwnedTrip(tripId, ownerId)
        val image = imageRepository.findById(imageId).orElseThrow { ServiceException(TripErrorCode.IMAGE_NOT_FOUND) }
        if (image.trip.id != trip.id || image.owner.id != ownerId) throw ServiceException(TripErrorCode.IMAGE_FORBIDDEN)
        trip.changeRepresentativeImage(image)
        return TripResponse(trip)
    }

    @Transactional(readOnly = true)
    fun findTop10PublicTripsByLikeCount(): List<TripResponse> =
        tripRepository.findTop10PublicTripsByRecentLikeCount(LocalDateTime.now().minusMonths(1)).map(::TripResponse)

    @Transactional(readOnly = true)
    fun findWeeklyTrendingTrips(): List<WeeklyTrendingTripResponse> =
        tripRepository.findWeeklyTrendingTrips(LocalDateTime.now().minusDays(7), PageRequest.of(0, 9)).map {
            WeeklyTrendingTripResponse(TripResponse(it[0] as Trip), (it[1] as Number).toLong())
        }

    @Transactional(readOnly = true)
    fun findPopularDestinations(): List<PopularDestinationResponse> =
        tripRepository.findPopularDestinations(PageRequest.of(0, 6)).map { row ->
            val country = (row[0] as String).trim()
            val city = (row[1] as String).trim()
            val representative =
                tripRepository
                    .findFirstByVisibilityTrueAndCountryAndCityOrderByLikeCountDescCreatedAtDescIdDesc(
                        row[0] as String,
                        row[1] as String,
                    ).orElse(null)
            PopularDestinationResponse(
                country,
                city,
                (row[2] as Number).toLong(),
                representative?.representativeImage?.thumbnailUrl,
                representative?.id,
            )
        }

    @Transactional(readOnly = true)
    fun findPublicTripsByCreatedAtDesc(): List<TripResponse> = tripRepository.findByVisibilityTrueOrderByCreatedAtDesc().map(::TripResponse)

    @Transactional(readOnly = true)
    fun findPublicTripsByCreatedAtDesc(pageable: Pageable): Page<TripResponse> =
        tripRepository.findByVisibilityTrueOrderByCreatedAtDescIdDesc(pageable).map(::TripResponse)

    private fun findTrip(tripId: Long): Trip = tripRepository.findById(tripId).orElseThrow { ServiceException(TripErrorCode.NOT_FOUND) }

    private fun validateOwner(
        trip: Trip,
        ownerId: Long?,
    ) {
        if (trip.owner.id != ownerId) throw ServiceException(TripErrorCode.FORBIDDEN)
    }
}
